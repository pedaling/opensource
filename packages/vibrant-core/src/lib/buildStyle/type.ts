import type { CurrentTheme } from '@vibrant-ui/theme';

type StyleObject = {
  [property: string]: any;
};

export type BuildStyleFn = (styleObjects: StyleObject[], info: { theme?: CurrentTheme }) => Record<string, any>;
