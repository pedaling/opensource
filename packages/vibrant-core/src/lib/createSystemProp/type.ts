import type { Theme } from '@vibrant-ui/theme';
import type { Path } from '@vibrant-ui/utils';

export type SystemPropConfig = {
  property: string;
  styleProperty?: string;
  scale?: Extract<Path<Theme>, 'colors' | 'opacity' | 'typography' | 'typographyWeight'>;
  disabled?: boolean;
  transform?: (value: any) => Record<string, any> | undefined;
};

export type SystemProp = {
  (props: Record<string, any>): Record<string, any>;
  propName: string;
  disabled: boolean;
};
