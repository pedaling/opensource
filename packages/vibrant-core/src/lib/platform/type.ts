export type Platform = 'android' | 'ios' | 'macos' | 'web' | 'windows';
