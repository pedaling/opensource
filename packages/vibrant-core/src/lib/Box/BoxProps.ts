import type { ComponentType, Ref } from 'react';
import type { DistributiveOmit, LayoutEvent } from '@vibrant-ui/utils';
import type { ReactElementChildren } from '../../types';
import { createInterpolation } from '../createInterpolation';
import { createShouldForwardProp } from '../createShouldForwardProp';
import { injectContext } from '../injectContext';
import type {
  BackgroundSystemProps,
  BorderSystemProps,
  ColorSystemProps,
  DisplaySystemProps,
  ElevationSystemProps,
  FlexboxSystemProps,
  InputSystemProps,
  InteractionSystemProps,
  OverflowSystemProps,
  PositionSystemProps,
  PseudoClassSystemProps,
  SizingSystemProps,
  SpacingSystemProps,
  SvgSystemProps,
  TextSystemProps,
  TransformSystemProps,
  TypographySystemProps,
} from '../props';
import {
  backgroundSystemProps,
  borderSystemProps,
  colorSystemProps,
  displaySystemProps,
  elevationSystemProps,
  flexboxSystemProps,
  inputSystemProps,
  interactionSystemProps,
  overflowSystemProps,
  positionSystemProps,
  pseudoClassSystemProps,
  sizingSystemProps,
  spacingSystemProps,
  svgSystemProps,
  textSystemProps,
  transformSystemProps,
  typographySystemProps,
} from '../props';

export type SystemProps = BackgroundSystemProps &
  BorderSystemProps &
  ColorSystemProps &
  DisplaySystemProps &
  ElevationSystemProps &
  FlexboxSystemProps &
  InputSystemProps &
  InteractionSystemProps &
  OverflowSystemProps &
  PositionSystemProps &
  PseudoClassSystemProps &
  SizingSystemProps &
  SpacingSystemProps &
  SvgSystemProps &
  TextSystemProps &
  TypographySystemProps &
  TransformSystemProps;

const systemProps = [
  ...backgroundSystemProps,
  ...borderSystemProps,
  ...colorSystemProps,
  ...displaySystemProps,
  ...elevationSystemProps,
  ...flexboxSystemProps,
  ...inputSystemProps,
  ...interactionSystemProps,
  ...overflowSystemProps,
  ...positionSystemProps,
  ...pseudoClassSystemProps,
  ...sizingSystemProps,
  ...spacingSystemProps,
  ...svgSystemProps,
  ...textSystemProps,
  ...typographySystemProps,
  ...transformSystemProps,
];

const systemPropNames = systemProps.filter(systemProp => !systemProp.disabled).map(systemProp => systemProp.propName);

export type BoxElements =
  | 'article'
  | 'aside'
  | 'button'
  | 'clipPath'
  | 'defs'
  | 'div'
  | 'footer'
  | 'g'
  | 'header'
  | 'img'
  | 'input'
  | 'label'
  | 'li'
  | 'linearGradient'
  | 'main'
  | 'mask'
  | 'nav'
  | 'path'
  | 'section'
  | 'span'
  | 'stop'
  | 'svg'
  | 'ul';

type ComponentProps<Component extends ComponentType | undefined> = Component extends undefined
  ? undefined
  : Component extends ComponentType<infer ComponentProps>
  ? ComponentProps
  : undefined;

export type BoxProps<
  BaseComponent extends ComponentType | undefined = undefined,
  ElementName extends BoxElements | undefined = undefined
> = SystemProps & {
  as?: ElementName;
  base?: BaseComponent;
  id?: string;
  onLayout?: (layoutEvent: LayoutEvent) => void;
} & DistributiveOmit<
    ComponentProps<BaseComponent> extends undefined
      ? (ElementName extends keyof JSX.IntrinsicElements
          ? JSX.IntrinsicElements[ElementName]
          : Record<never, never>) & {
          ref?: Ref<HTMLElement>;
          children?: ReactElementChildren;
        }
      : ComponentProps<BaseComponent>,
    keyof SystemProps
  >;

export type { LayoutEvent };

export const interpolation = injectContext(
  createInterpolation(systemProps, { display: 'flex', boxSizing: 'border-box', position: 'relative' })
);

export const shouldForwardProp = createShouldForwardProp(systemPropNames);
