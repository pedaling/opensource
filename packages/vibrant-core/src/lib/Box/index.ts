export type { BoxProps, LayoutEvent } from './BoxProps';
export { Box } from './Box';
