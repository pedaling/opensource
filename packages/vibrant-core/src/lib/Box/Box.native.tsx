import type { ComponentType } from 'react';
import { forwardRef, memo } from 'react';
import type { LayoutChangeEvent } from 'react-native';
import { StyleSheet, TextInput, TouchableOpacity, View } from 'react-native';
import styled from '@emotion/native';
import { isDefined } from '@vibrant-ui/utils';
import { getElementPosition } from '../getElementPosition';
import { OnColorContainer } from '../OnColorContainer';
import type { BoxProps } from './BoxProps';
import { interpolation, shouldForwardProp } from './BoxProps';

const transformAs = (as: keyof JSX.IntrinsicElements): ComponentType => {
  switch (as) {
    case 'button':
      return TouchableOpacity;
    case 'input':
      return TextInput;
    default:
      return View;
  }
};

export const Box = memo(
  styled(
    forwardRef<HTMLDivElement, BoxProps & { style: any }>(
      (
        {
          as,
          base,
          style,
          onLayout,
          backgroundColor,
          id,
          role,
          ariaLabel,
          ariaLabelledBy,
          ariaChecked,
          ariaSelected,
          ariaCurrent: _ariaCurrent,
          ...restProps
        },
        ref
      ) => {
        const { BaseComponent, props, ...restStyle } = StyleSheet.flatten(style);

        const Component = BaseComponent ?? base ?? transformAs(as ?? 'div');

        const handleLayout = onLayout
          ? async (event: LayoutChangeEvent) => {
              const layout = await getElementPosition(event.currentTarget);

              onLayout(layout);
            }
          : undefined;

        const element = (
          <Component
            ref={ref}
            id={id}
            role={role}
            style={restStyle}
            onLayout={handleLayout}
            accessibilityLabel={ariaLabel}
            accessibilityLabelledBy={ariaLabelledBy}
            {...(isDefined(ariaChecked) || isDefined(ariaSelected)
              ? { accessibilityState: { checked: ariaChecked, selected: ariaSelected } }
              : {})}
            collapsable={ref ? false : undefined}
            {...(base ? { as } : {})}
            {...restProps}
            {...props}
          />
        );

        if (backgroundColor) {
          return <OnColorContainer backgroundColor={backgroundColor}>{element}</OnColorContainer>;
        }

        return element;
      }
    ),
    {
      shouldForwardProp,
    }
  )(interpolation)
);

Box.displayName = 'Box';
