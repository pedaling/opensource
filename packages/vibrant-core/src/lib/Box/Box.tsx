import type { ComponentType, FC, ReactElement } from 'react';
import { createElement, forwardRef, useCallback } from 'react';
import styled from '@emotion/styled';
import { useLayoutEvent } from '@vibrant-ui/utils';
import { OnColorContainer } from '../OnColorContainer';
import type { BoxElements, BoxProps } from './BoxProps';
import { interpolation, shouldForwardProp } from './BoxProps';

export const Box = styled(
  forwardRef<HTMLDivElement, BoxProps>(({ as, base, onLayout, ariaLabel, backgroundColor, ...restProps }, ref) => {
    const Component = base ? (base as ComponentType<any>) : undefined;

    const { ref: layoutEventRef } = useLayoutEvent(onLayout);

    const composeRef = useCallback(
      (node: HTMLDivElement) => {
        if (ref) {
          if (typeof ref === 'function') {
            ref(node);
          } else {
            ref.current = node;
          }
        }

        layoutEventRef(node);
      },
      [ref, layoutEventRef]
    );

    if (Component) {
      return (
        <OnColorContainer backgroundColor={backgroundColor}>
          <Component ref={composeRef} {...(as ? { as } : {})} aria-label={ariaLabel} {...restProps} />
        </OnColorContainer>
      );
    }

    return (
      <OnColorContainer backgroundColor={backgroundColor}>
        {createElement(as ?? 'div', {
          ref: composeRef,
          'aria-label': ariaLabel,
          ...restProps,
        })}
      </OnColorContainer>
    );
  }),
  {
    shouldForwardProp,
  }
)(interpolation) as <
  BaseComponent extends ComponentType<any> | undefined = undefined,
  ElementName extends BoxElements | undefined = undefined
>(
  props: BoxProps<BaseComponent, ElementName>
) => ReactElement;

(Box as FC).displayName = 'Box';
