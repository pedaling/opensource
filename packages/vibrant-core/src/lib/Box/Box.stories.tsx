import type { ComponentStory, ComponentMeta } from '@storybook/react';
import { Box } from './Box';

export default {
  title: 'Box',
  component: Box,
  args: {
    width: 300,
    height: 300,
    backgroundColor: 'primary',
  },
} as ComponentMeta<typeof Box>;

export const Basic: ComponentStory<typeof Box> = props => <Box {...props}>Box</Box>;
