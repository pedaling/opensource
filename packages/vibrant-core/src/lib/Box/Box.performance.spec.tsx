import React, { Profiler } from 'react';
import styled from '@emotion/styled';
import type { ReactRenderer } from '@vibrant-ui/utils/testing-web';
import { createReactRenderer } from '@vibrant-ui/utils/testing-web';
import { Box } from './Box';

describe('<Box /> Performance', () => {
  const { render } = createReactRenderer();
  let vibrantRenderer: ReactRenderer;
  let emotionRenderer: ReactRenderer;

  const createOnRenderCallback = () => {
    let totalDuration = 0;
    let renderCount = 0;
    const onRender = (
      _id: string,
      _phase: 'mount' | 'update',
      actualDuration: number,
      _baseDuration: number,
      _startTime: number,
      _commitTime: number,
      _interactions: Set<any>
    ) => {
      totalDuration += actualDuration;
      renderCount++;
    };

    return { onRender, getAverageDuration: () => (renderCount ? totalDuration / renderCount : 0) };
  };

  const EmotionBoxComponent = () => {
    const StyledComponent = styled.div`
      display: flex;
      align-content: stretch;
      flex-shrink: 1;
      flex-direction: column;
      box-sizing: border-box;
      position: relative;
      width: 300px;
      height: 300px;
      background-color: red;
    `;

    return <StyledComponent />;
  };

  describe('Mount Performance', () => {
    beforeEach(() => {
      vibrantRenderer = render(
        <Profiler id="vibrant" onRender={createOnRenderCallback().onRender}>
          <Box width={300} height={300} backgroundColor="red" />
        </Profiler>
      );

      emotionRenderer = render(
        <Profiler id="emotion" onRender={createOnRenderCallback().onRender}>
          <EmotionBoxComponent />
        </Profiler>
      );
    });

    it('div element created', () => {
      expect(vibrantRenderer.container.querySelector('div')).toBeTruthy();
      expect(emotionRenderer.container.querySelector('div')).toBeTruthy();
    });

    it('match snapshot', () => {
      expect(vibrantRenderer.container).toMatchSnapshot('vibrant box snapshot');
      expect(emotionRenderer.container).toMatchSnapshot('emotion box snapshot');
    });

    it('should mount <Box /> quickly over multiple iterations', () => {
      const iterations = 1000;
      let vibrantTotal = 0;
      let emotionTotal = 0;

      for (let i = 0; i < iterations; i++) {
        const vibrantCallback = createOnRenderCallback();
        const emotionCallback = createOnRenderCallback();

        vibrantRenderer = render(
          <Profiler id={`vibrant-${i}`} onRender={vibrantCallback.onRender}>
            <Box width={300} height={300} backgroundColor="red" />
          </Profiler>
        );

        emotionRenderer = render(
          <Profiler id={`emotion-${i}`} onRender={emotionCallback.onRender}>
            <EmotionBoxComponent />
          </Profiler>
        );

        vibrantTotal += vibrantCallback.getAverageDuration();
        emotionTotal += emotionCallback.getAverageDuration();
      }

      const vibrantAvg = vibrantTotal / iterations;
      const emotionAvg = emotionTotal / iterations;

      expect(vibrantAvg / emotionAvg).toBeLessThan(3);
    });
  });
});
