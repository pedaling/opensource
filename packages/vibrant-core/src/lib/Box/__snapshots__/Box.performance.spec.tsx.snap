// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Box /> Performance Mount Performance match snapshot: emotion box snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
  width: 300px;
  height: 300px;
  background-color: red;
}

<div>
  <div
    class="emotion-0"
  />
</div>
`;

exports[`<Box /> Performance Mount Performance match snapshot: vibrant box snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
  width: 300px;
  height: 300px;
  background-color: red;
}

<div>
  <div
    class="emotion-0"
  />
</div>
`;
